File: python/leetcode/editor/cn/[652]寻找重复的子树.py
from typing import *
from util import *


# leetcode submit region begin(Prohibit modification and deletion)
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def __init__(self):
        self.subTrees = {}
        self.res = []

    def findDuplicateSubtrees(self, root: Optional[TreeNode]) -> List[Optional[TreeNode]]:
        self.serialize(root)
        return self.res

    def serialize(self, root: Optional[TreeNode]):
        if root is None:
            return "#"

        # 先计算左右子树的序列化结果
        left = self.serialize(root.left)
        right = self.serialize(root.right)

        mySelf = f"{left},{right},{root.val}"
        freq = self.subTrees.setdefault(mySelf, 0)
        # 多次重复也只会被加入结果一次
        if freq == 1:
            self.res.append(root)

        self.subTrees[mySelf] = freq + 1
        return mySelf

# leetcode submit region end(Prohibit modification and deletion)

File: cpp/leetcodecn/[652]寻找重复的子树.cpp
#include <vector>
#include <sstream>
#include "util.h"
#include <map>
#include <stack>
#include <string>
#include <utility>
#include <memory>
#include <iterator>
#include <unordered_map>
#include <algorithm>
#include <numeric>

using namespace std;
//leetcode submit region begin(Prohibit modification and deletion)
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
public:
    vector<TreeNode *> findDuplicateSubtrees(TreeNode *root) {
        serialize(root);
        return res;
    }

private:
    unordered_map<string, int> subTrees;
    vector<TreeNode *> res;

    string serialize(TreeNode *node) {
        if (node == nullptr) return "#";

        auto left = serialize(node->left);
        auto right = serialize(node->right);
        auto mySelf = left + "," + right + "," + to_string(node->val);

        auto freq = subTrees[mySelf];
        if (freq == 1) res.push_back(node);

        subTrees[mySelf] = freq + 1;

        return mySelf;
    }
};
//leetcode submit region end(Prohibit modification and deletion)
